jlm/rvsdg/FunctionType.hpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 Nico Reißmann <nico.reissmann@gmail.com>
+ * Copyright 2025 Helge Bahmann <hcb@chaoticmind.net>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_FUNCTION_TYPE_HPP
+#define JLM_RVSDG_FUNCTION_TYPE_HPP
+
+#include <jlm/rvsdg/type.hpp>
+#include <jlm/util/common.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+/**
+ * \brief Function type class
+ *
+ * Represents the type of a callable function.
+ */
+class FunctionType final : public ValueType
+{
+public:
+  ~FunctionType() noexcept override;
+
+  FunctionType(
+      std::vector<std::shared_ptr<const jlm::rvsdg::Type>> argumentTypes,
+      std::vector<std::shared_ptr<const jlm::rvsdg::Type>> resultTypes);
+
+  inline const std::vector<std::shared_ptr<const jlm::rvsdg::Type>> &
+  Arguments() const noexcept
+  {
+    return ArgumentTypes_;
+  }
+
+  const std::vector<std::shared_ptr<const jlm::rvsdg::Type>> &
+  Results() const noexcept
+  {
+    return ResultTypes_;
+  }
+
+  size_t
+  NumResults() const noexcept
+  {
+    return ResultTypes_.size();
+  }
+
+  size_t
```

Comment:
nodiscard

Reply:
I am BTW inclined to remove all those convenience accosser members -- just exposing Results() is enough, caller can just apply "size" to that etc.